src/Functions/dateDiff.cpp

```diff
@@ -33,6 +33,199 @@ namespace ErrorCodes
 namespace
 {
 
+struct ToExtendedRelativeYearNumImpl
```

Comment:
I agree this approach introduces less "code bloating". I probably like it more than current approach. While it makes extended variants of toRelative() to be some kind of public functions (while there are not strictly public because they are protected by ExtendedPrecision), while in my current code they are private - they don't go outside of dateDiff.cpp.

So I will rewrite my code to use your approach. 

BTW what about naming of the ExtendedPrecision? It's not about precision but it's about extended result. Maybe ExtendedResult is a better name? 

Also maybe make it `false` by default? 
` template<bool ExtendedPrecision = false>`

Reply:
Also in previous PR I implemented:
src/Functions/DateTimeTransforms.h

```
template <typename FromDataType, typename ToDataType, typename Transform, bool is_extended_result = false>
struct DateTimeTransformImpl
```

In CH guides the variable naming should be in extended_precision and not ExtendedPrecision. 

So maybe simply use is_extended_result  for toRelative functions.

Reply:
Implemented it the way described above and pushed. 
While use it this way:
```
template<bool is_extended_result>
struct ToRelativeMonthNumImpl
```
So use it without default value. Strange, with default value:
```
template<bool is_extended_result = false>
struct ToRelativeMonthNumImpl
```

It does not compile:

> toRelativeMonthNum.cpp:10:86: error: use of class template 'ToRelativeMonthNumImpl' requires template arguments
> [build] using FunctionToRelativeMonthNum = FunctionDateOrDateTimeToSomething<DataTypeUInt32, ToRelativeMonthNumImpl>;
> 


In ToRelativeHourNumImpl please see:

```
    static inline UInt32 execute(UInt32 t, const DateLUTImpl & time_zone)
    {
        if constexpr (is_extended_result)
            return time_zone.toStableRelativeHourNum(static_cast<time_t>(t));
        else
            return time_zone.toRelativeHourNum(static_cast<time_t>(t));
    }
    // ... skip ...
    static inline UInt32 execute(UInt16 d, const DateLUTImpl & time_zone)
    {
        if constexpr (is_extended_result)
            return time_zone.toStableRelativeHourNum(DayNum(d));
        else
            return time_zone.toRelativeHourNum(DayNum(d));
    }
```
So usage of ``is_extended_result`` is not strictly extended. Because the result type is the same, but in this case it's rather a compatibility handling. 
It's needed to do calculations "via" Unix epoch. I will explained in in another thread of this PR.

Reply:
> BTW what about naming of the ExtendedPrecision? It's not about precision but it's about extended result. Maybe ExtendedResult is a better name?

Agree, `ExtendedResult` sounds better. Didn't put much thought into naming.

> Also maybe make it false by default?
> `template<bool ExtendedPrecision = false>`

I have witnessed too many unnecessary disasters with default function parameters or default template parameters. See e.g. #37117 for a recent example in ClickHouse which led to wrong results. My personal preference would be to make stuff explicit, unless it is a ton of work.

Reply:
Totally optional, but if you like to make this even more fancy, you could use a scoped enum as template parameter

```cpp
enum class ResultPrecision
{
    Standard,
    Extended
};

// in the template:
template <ResultPrecision precision_>
class ToRelativeYearNumImpl

   if constexpr(precision_ == ResultPrecision::Standard)
   {
        ...
   }

// the main advantage is that the template instantiation becomes prettier:

ToRelativeYearNumImpl<ResultPrecision::Standard> impl;
```

Reply:
> It does not compile:

>     toRelativeMonthNum.cpp:10:86: error: use of class template 'ToRelativeMonthNumImpl' requires template arguments
>    [build] using FunctionToRelativeMonthNum = FunctionDateOrDateTimeToSomething<DataTypeUInt32, ToRelativeMonthNumImpl>;

I think the syntax to use a default template parameter would be `ToRelativeMonthNumImpl<>`.

Reply:
I re-wrote the code to use ResultPrecision::Standard. Please take a look. Yes, it's prettier and it's more easy to read / understand such code.